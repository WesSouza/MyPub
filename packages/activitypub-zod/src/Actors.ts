import { z } from "zod";

import { DateValue, dateValue, UrlValue, urlValue } from "./common.js";
import { lazyObjectBaseSchema, ObjectBase } from "./Objects.js";

export type Actor = Omit<ObjectBase, "id" | "type"> & {
  id: string;
  type: "Application" | "Group" | "Organization" | "Person" | "Service";
  inbox?: UrlValue | null | undefined;
  outbox?: UrlValue | null | undefined;
  endpoints?:
    | {
        sharedInbox?: UrlValue | null | undefined;
      }
    | null
    | undefined;
  followers?: UrlValue | null | undefined;
  following?: UrlValue | null | undefined;
  preferredUsername?: string | null | undefined;

  // Mastodon
  devices?: UrlValue | null | undefined;
  discoverable?: boolean | null | undefined;
  featured?: UrlValue | null | undefined;
  featuredTags?: UrlValue | null | undefined;
  manuallyApprovesFollowers?: boolean | null | undefined;
  published?: DateValue | null | undefined;

  publicKey?:
    | {
        id: string;
        owner: string;
        publicKeyPem: string;
      }
    | null
    | undefined;
};

export const lazyActor = () =>
  lazyObjectBaseSchema()
    .omit({ type: true })
    .extend({
      id: z.string(),
      type: z.enum([
        "Application",
        "Group",
        "Organization",
        "Person",
        "Service",
      ]),

      inbox: urlValue,
      outbox: urlValue,
      endpoints: z
        .object({
          sharedInbox: urlValue.nullish(),
        })
        .nullish(),
      followers: urlValue.nullish(),
      following: urlValue.nullish(),
      preferredUsername: z.string().nullish(),

      // Mastodon
      devices: urlValue.nullish(),
      discoverable: z.boolean().nullish(),
      featured: urlValue.nullish(),
      featuredTags: urlValue.nullish(),
      manuallyApprovesFollowers: z.boolean().nullish(),
      published: dateValue.nullish(),

      // Security
      publicKey: z
        .object({
          id: z.string(),
          owner: z.string(),
          publicKeyPem: z.string(),
        })
        .nullish(),
    });

export const ActorSchema: z.ZodType<Actor> = z.lazy(lazyActor);
