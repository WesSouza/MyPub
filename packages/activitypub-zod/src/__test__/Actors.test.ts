import { readFile } from "fs/promises";
import { describe, expect, it } from "vitest";

import { ActorSchema } from "../Actors.js";

describe("Actors", () => {
  it("parses https://mastodon.social/users/gargron", async () => {
    const json = await readFile(
      new URL("./fixtures/mastodon-social-user-gargron.json", import.meta.url),
      "utf-8",
    );
    const data = ActorSchema.parse(JSON.parse(json));
    expect(data).toMatchSnapshot();
  });
});
