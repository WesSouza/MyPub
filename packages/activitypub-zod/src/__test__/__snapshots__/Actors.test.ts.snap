// Vitest Snapshot v1

exports[`Actors > parses https://mastodon.social/users/gargron 1`] = `
{
  "@context": [
    "https://www.w3.org/ns/activitystreams",
    "https://w3id.org/security/v1",
    {
      "Curve25519Key": "toot:Curve25519Key",
      "Device": "toot:Device",
      "Ed25519Key": "toot:Ed25519Key",
      "Ed25519Signature": "toot:Ed25519Signature",
      "EncryptedMessage": "toot:EncryptedMessage",
      "PropertyValue": "schema:PropertyValue",
      "alsoKnownAs": {
        "@id": "as:alsoKnownAs",
        "@type": "@id",
      },
      "cipherText": "toot:cipherText",
      "claim": {
        "@id": "toot:claim",
        "@type": "@id",
      },
      "deviceId": "toot:deviceId",
      "devices": {
        "@id": "toot:devices",
        "@type": "@id",
      },
      "discoverable": "toot:discoverable",
      "featured": {
        "@id": "toot:featured",
        "@type": "@id",
      },
      "featuredTags": {
        "@id": "toot:featuredTags",
        "@type": "@id",
      },
      "fingerprintKey": {
        "@id": "toot:fingerprintKey",
        "@type": "@id",
      },
      "focalPoint": {
        "@container": "@list",
        "@id": "toot:focalPoint",
      },
      "identityKey": {
        "@id": "toot:identityKey",
        "@type": "@id",
      },
      "manuallyApprovesFollowers": "as:manuallyApprovesFollowers",
      "messageFranking": "toot:messageFranking",
      "messageType": "toot:messageType",
      "movedTo": {
        "@id": "as:movedTo",
        "@type": "@id",
      },
      "publicKeyBase64": "toot:publicKeyBase64",
      "schema": "http://schema.org#",
      "suspended": "toot:suspended",
      "toot": "http://joinmastodon.org/ns#",
      "value": "schema:value",
    },
  ],
  "attachment": [
    {
      "name": "Patreon",
      "type": "PropertyValue",
      "value": "<a href=\\"https://www.patreon.com/mastodon\\" target=\\"_blank\\" rel=\\"nofollow noopener noreferrer me\\"><span class=\\"invisible\\">https://www.</span><span class=\\"\\">patreon.com/mastodon</span><span class=\\"invisible\\"></span></a>",
    },
  ],
  "devices": "https://mastodon.social/users/Gargron/collections/devices",
  "discoverable": true,
  "endpoints": {
    "sharedInbox": "https://mastodon.social/inbox",
  },
  "featured": "https://mastodon.social/users/Gargron/collections/featured",
  "featuredTags": "https://mastodon.social/users/Gargron/collections/tags",
  "followers": "https://mastodon.social/users/Gargron/followers",
  "following": "https://mastodon.social/users/Gargron/following",
  "icon": {
    "mediaType": "image/jpeg",
    "type": "Image",
    "url": "https://files.mastodon.social/accounts/avatars/000/000/001/original/dc4286ceb8fab734.jpg",
  },
  "id": "https://mastodon.social/users/Gargron",
  "image": {
    "mediaType": "image/jpeg",
    "type": "Image",
    "url": "https://files.mastodon.social/accounts/headers/000/000/001/original/3b91c9965d00888b.jpeg",
  },
  "inbox": "https://mastodon.social/users/Gargron/inbox",
  "manuallyApprovesFollowers": false,
  "name": "Eugen Rochko",
  "outbox": "https://mastodon.social/users/Gargron/outbox",
  "preferredUsername": "Gargron",
  "publicKey": {
    "id": "https://mastodon.social/users/Gargron#main-key",
    "owner": "https://mastodon.social/users/Gargron",
    "publicKeyPem": "-----BEGIN PUBLIC KEY-----
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAvXc4vkECU2/CeuSo1wtn
Foim94Ne1jBMYxTZ9wm2YTdJq1oiZKif06I2fOqDzY/4q/S9uccrE9Bkajv1dnkO
Vm31QjWlhVpSKynVxEWjVBO5Ienue8gND0xvHIuXf87o61poqjEoepvsQFElA5ym
ovljWGSA/jpj7ozygUZhCXtaS2W5AD5tnBQUpcO0lhItYPYTjnmzcc4y2NbJV8hz
2s2G8qKv8fyimE23gY1XrPJg+cRF+g4PqFXujjlJ7MihD9oqtLGxbu7o1cifTn3x
BfIdPythWu5b4cujNsB3m3awJjVmx+MHQ9SugkSIYXV0Ina77cTNS0M2PYiH1PFR
TwIDAQAB
-----END PUBLIC KEY-----
",
  },
  "published": "2016-03-16T00:00:00Z",
  "summary": "<p>Founder, CEO and lead developer <span class=\\"h-card\\"><a href=\\"https://mastodon.social/@Mastodon\\" class=\\"u-url mention\\">@<span>Mastodon</span></a></span>, Germany.</p>",
  "tag": [],
  "type": "Person",
  "url": "https://mastodon.social/@Gargron",
}
`;
